src/main/java/seedu/tasklist/model/ModelManager.java

```diff
@@ -73,20 +83,50 @@ public synchronized void deleteTask(ReadOnlyTask target) throws TaskNotFoundExce
         taskList.removeTask(target);
         updateFilteredListToShowIncomplete();
         indicateTaskListChanged();
+        addToUndoStack(UndoCommand.DEL_CMD_ID, (Task)target);
     }
 
     @Override
     public synchronized void addTask(Task task) throws UniqueTaskList.DuplicateTaskException {
         taskList.addTask(task);
         updateFilteredListToShowIncomplete();
         indicateTaskListChanged();
+        addToUndoStack(UndoCommand.ADD_CMD_ID, task);
```

Comment:
we need 1 new method specifically for undoAdd that calls addTask, and in that method you addToUndoStack. Same for all other list operations (delete, update, mark). Otherwise undos end up pushing to the undostack as well.